File: Wirtualny_Swiat_Java/Projekt_PO_Java/src/model/Organism.java
package model;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public abstract class Organism {
    private final int initiative;
    private int age, power;
    private boolean alive = true;
    protected World world;
    Point position;
    public final Random random = new Random();

    public Organism(World world, Point position, int power, int initiative){
        this.world = world;
        this.position = position;
        this.initiative = initiative;
        this.power = power;
    }

    public abstract Color getColor();
    public abstract void Action();
    public abstract String getNazwa();
    public abstract void Collision(Organism attacker);
    public abstract Organism makeNewOrganism(Point position);

    public Point getPosition() {
        return position;
    }

    public void setPosition(Point position) {
        this.position = position;
    }

    public void setAlive(boolean alive) {
        this.alive = alive;
    }

    public boolean getIsAlive() {
        return alive;
    }

    public int getInitiative() {
        return initiative;
    }

    public int getAge() {
        return age;
    }

    public int getPower() {
        return power;
    }

    public void setPower(int power){
        this.power = power;
    }

    protected void increaseAge(){
        this.age += 1;
    }

    protected Point findPosition() {
        List<Point> positions = new ArrayList<>();
        Point currentPosition = getPosition();

        // Sprawdź wszystkie 4 sąsiednie pola
        int x = currentPosition.getX();
        int y = currentPosition.getY();

        Point[] directions = new Point[] {
                new Point(x, y - 1), // góra
                new Point(x + 1, y), // prawo
                new Point(x, y + 1), // dół
                new Point(x - 1, y)  // lewo
        };

        for (Point pos : directions) {
            if (world.isCorrectPosition(pos) && world.isEmpty(pos)) {
                positions.add(pos);
            }
        }

        if (positions.isEmpty()) {
            return currentPosition; // nie ma gdzie się rozmnożyć
        }

        return positions.get(random.nextInt(positions.size())); // losowa dostępna pozycja
    }

    public void collisionHelper(Organism attacker, Point position){
        if(attacker.getPower() >= this.getPower()){
            this.setAlive(false);
            world.addEvent(attacker.getNazwa() + " zabił " + this.getNazwa() + " na pozycji (" + position.getX() + ", " + position.getY() + ")");
            if (attacker.getIsAlive()){
                world.moveOrganism(attacker, position);
            }

        }else {
            attacker.setAlive(false);
            world.addEvent(this.getNazwa() + " zaił " + attacker.getNazwa() + " na pozycji (" + position.getX() + ", " + position.getY() + ")");
            if (this.getIsAlive()){
                world.moveOrganism(this, position);
            }
        }
    }

}

File: WirtualnySwiatPython/Organism.py
import random
from abc import ABC, abstractmethod

from point import Point


class Organism(ABC):
    def __init__(self, world, position, power, initiative):
        self.world = world
        self.position = position
        self.power = power
        self.initiative = initiative
        self.isAlive = True
        self.age = 0

    def getPosition(self):
        return self.position

    def setPosition(self, pos):
        self.position = pos

    def getIsAlive(self):
        return self.isAlive

    def setIsAlive(self, tmp):
        self.isAlive = tmp

    def getPower(self):
        return self.power

    def setPower(self, tmp):
        self.power = tmp

    def getInitiative(self):
        return self.initiative

    def setInitiative(self, tmp):
        self.initiative = tmp

    def setAge(self, tmp):
        self.age = tmp

    def getAge(self):
        return self.age

    @abstractmethod
    def Action(self):
        pass
    @abstractmethod
    def getColor(self):
        pass
    @abstractmethod
    def getNazwa(self):
        pass

    @abstractmethod
    def makeNewOrganism(self, position):
        pass

    def increaseAge(self):
        self.age += 1

    def findPosition(self):
        positions = []
        current = self.getPosition()

        x = current.GetX()
        y = current.GetY()

        directions = [
            Point(x, y -1),
            Point(x + 1, y),
            Point(x, y + 1),
            Point(x - 1, y)
        ]

        for dir in directions:
            if self.world.isCorrectPosition(dir) and self.world.isEmptyPosition(dir):
                positions.append(dir)

        if not positions:
            return None

        return random.choice(positions)

    def collisionHelper(self, attacker, position):
        if attacker.getPower() >= self.getPower():
            self.setIsAlive(False)
            self.world.addEvent(
                f"{attacker.getNazwa()} zabił {self.getNazwa()} na pozycji ({position.GetX()}, {position.GetY()})")

            if attacker.getIsAlive():
                self.world.moveOrganism(attacker, position)

        else:
            from Animals.Human import Human
            if isinstance(attacker, Human):
                self.world.killHuman()
                return

            attacker.setIsAlive(False)
            self.world.addEvent(
                f"{self.getNazwa()} zabił {attacker.getNazwa()} na pozycji ({position.GetX()}, {position.GetY()})")

            if self.getIsAlive():
                self.world.moveOrganism(self, position)

